xarray/core/groupby.py

```diff
@@ -367,35 +394,64 @@ def __post_init__(self) -> None:
             self.inserted_dims,
         ) = _ensure_1d(group=self.group, obj=self.obj)
 
+        self.factorize()
+
     @property
     def name(self) -> Hashable:
-        return self.group1d.name
+        # the name has to come from unique_coord because we need `_bins` suffix for BinGrouper
+        return self.unique_coord.name
 
     @property
     def size(self) -> int:
         return len(self)
 
     def __len__(self) -> int:
-        return len(self.full_index)  # TODO: full_index not def, abstractmethod?
+        return len(self.full_index)
 
     @property
     def dims(self):
         return self.group1d.dims
 
-    @abstractmethod
-    def factorize(self) -> T_FactorizeOut:
-        raise NotImplementedError
-
-    def _factorize(self) -> None:
-        # This design makes it clear  to mypy that
-        # codes, group_indices, unique_coord, and full_index
-        # are set by the factorize  method on the derived class.
+    def factorize(self) -> None:
         (
             self.codes,
             self.group_indices,
             self.unique_coord,
             self.full_index,
-        ) = self.factorize()
+        ) = self.grouper.factorize(self.group1d)
+
+
+class Grouper(ABC):
```

Comment:
```suggestion
class Grouper(ABC):
    """base Grouper class which allow specifying groupby instructions"""
```

(or a better formulation)